core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java

```diff
@@ -77,6 +78,13 @@ public static EnumerableCorrelate create(
   @Override public EnumerableCorrelate copy(RelTraitSet traitSet,
       RelNode left, RelNode right, CorrelationId correlationId,
       ImmutableBitSet requiredColumns, JoinRelType joinType) {
+    // Get new collations since the Correlate has new inputs and new condition.
+    ImmutableList<RelCollation> newCollations = ImmutableList.copyOf(
+            RelMdCollation.enumerableCorrelate(
+            getCluster().getMetadataQuery(), left, right, joinType));
+    if (!newCollations.isEmpty()) {
+      traitSet = traitSet.replace(newCollations);
```

Comment:
I see what you are trying to do here but you are kinda violating the contract of `copy` (_Creates a copy of this relational expression, perhaps changing traits and inputs_); the client of the method can no longer change traits (at least collation). Furthermore, if you do this inside `copy` you could call the `create` method directly which brings me to my initial comment. Instead of using copy inside the rule you could opt for the RelBuilderFactory.

Reply:
Maybe we're going at this the wrong way, and trying too hard. If you call `ProjectJoinTransposeRule` on a `MergeJoin` (which requires sorted input), and the `Project` does not, after being pushed down, give sorted output, then maybe the rule should just abort.

Just a thought.

Reply:
Julian, should we just skip matching for `MergeJoin` if the project does not sort output ?

Reply:
I got it, I thought too much when I tried to solve this problem. I understood that the other  enumerable join class should have right collations too considering the CALCITE-2554.  I will adjustment this PR that only update the collations for MergeJoin or abort the rule if the new inputs of MergeJoin are not sorted(the result of RelMdCollation.mergeJoin for new MergeJoin is  empty).

Is this the suitable solution?

Reply:
IMHO, `ProjectJoinTransposeRule` should just stop matching physical operators.

Reply:
Yes. Let's make ProjectJoinTransposeRule.INSTANCE just match LogicalJoin and LogicalProject. Add class arguments so that people can create other instances of the rule, if they wish.